File: 76.py
def longest_subarray_at_most_k_distinct(nums, k):
    from collections import defaultdict

    freq = defaultdict(int)
    left = 0
    distinct = 0
    max_len = 0

    for right, val in enumerate(nums):
        if freq[val] == 0:
            distinct += 1
        freq[val] += 1

        while distinct > k:
            left_val = nums[left]
            freq[left_val] -= 1
            if freq[left_val] == 0:
                distinct -= 1
            left += 1

        max_len = max(max_len, right - left + 1)

    return max_len

File: 76.js
function longestSubarrayAtMostKDistinct(nums, k) {
    const freq = new Map();  // value -> count
    let left = 0;
    let distinct = 0;
    let maxLen = 0;

    for (let right = 0; right < nums.length; right++) {
        const val = nums[right];
        if (!freq.has(val) || freq.get(val) === 0) {
            distinct++;
        }
        freq.set(val, (freq.get(val) || 0) + 1);

        while (distinct > k) {
            const leftVal = nums[left];
            freq.set(leftVal, freq.get(leftVal) - 1);
            if (freq.get(leftVal) === 0) {
                distinct--;
            }
            left++;
        }

        maxLen = Math.max(maxLen, right - left + 1);
    }

    return maxLen;
}
